pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,821 @@
+		reasonText := ""
+		if update.reason != nil && *update.reason != "" {
+			reasonText = fmt.Sprintf(" ( %s )", *update.reason)
+		}
+		glog.V(4).Infof("Updating build %s -> %s%s", buildDesc(build), *update.phase, reasonText)
+	}
+
+	// Ensure that a pod name annotation has been set on the build if a pod is available
+	if update.podNameAnnotation != nil && !common.HasBuildPodNameAnnotation(build) && pod != nil {
+		update.setPodNameAnnotation(pod.Name)
+	}
+
+	err := bc.patchBuild(build, update)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to update build %s with %v: %v", buildDesc(build), update, err))
+		return
+	}
+
+	if stateTransition && isTerminal(*update.phase) {
+		common.HandleBuildCompletion(build, bc.buildLister, bc.buildConfigGetter, bc.buildDeleter, bc.runPolicies)
+	}
+}
+
+func (bc *BuildController) patchBuild(build *buildapi.Build, update *buildUpdate) error {
+
+	// Create a patch using the buildUpdate object
+	updatedBuild, err := buildutil.BuildDeepCopy(build)
+	if err != nil {
+		return fmt.Errorf("cannot create a deep copy of build %s: %v", buildDesc(build), err)
+	}
+	update.apply(updatedBuild)
+	patch, err := validation.CreateBuildPatch(build, updatedBuild)
+	if err != nil {
+		return fmt.Errorf("failed to create a build patch: %v", err)
+	}
+
+	glog.V(5).Infof("Patching build %s with %v", buildDesc(build), update)
+	return bc.buildPatcher.Patch(build.Namespace, build.Name, patch)
+}
+
+func (bc *BuildController) handlePodNotFound(build *buildapi.Build) (*buildUpdate, error) {
+	// Make one last attempt to fetch the pod using the REST client
+	_, err := bc.podClient.Pods(build.Namespace).Get(buildapi.GetBuildPodName(build), metav1.GetOptions{})
+	if err == nil {
+		glog.V(2).Infof("Found missing pod for build %s by using direct client.", buildDesc(build))
+		return nil, nil
+	}
+	return podDeletedUpdate(), nil
+}
+
+func (bc *BuildController) getPodByKey(key string) (*kapi.Pod, error) {
+	obj, exists, err := bc.podInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		glog.V(2).Infof("Unable to retrieve pod %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return nil, err
+	}
+	if !exists {
+		glog.V(2).Infof("Pod %q has been deleted", key)
+		return nil, nil
+	}
+
+	return obj.(*kapi.Pod), nil
+}
+
+func (bc *BuildController) getBuildByKey(key string) (*buildapi.Build, error) {
+	obj, exists, err := bc.buildInformer.GetIndexer().GetByKey(key)
+	if err != nil {
+		glog.V(2).Infof("Unable to retrieve build %q from store: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return nil, err
+	}
+	if !exists {
+		glog.V(2).Infof("Build %q has been deleted", key)
+		return nil, nil
+	}
+
+	return obj.(*buildapi.Build), nil
+}
+
+func (bc *BuildController) podUpdated(old, cur interface{}) {
+	// A periodic relist will send update events for all known pods.
+	curPod := cur.(*kapi.Pod)
+	oldPod := old.(*kapi.Pod)
+	if curPod.ResourceVersion == oldPod.ResourceVersion {
```

Comment:
comment for why it's ok to ignore pod relists here so people don't copy this pattern.